src/server/scripts/EasternKingdoms/ScarletMonastery/boss_headless_horseman.cpp

```diff
@@ -154,218 +178,359 @@ std::vector<Position> const HeadlessHorsemanFlightPoints =
     { 1758.00f, 1367.00f, 19.51f }
 };
 
-std::vector<Position> const HeadlessHorsemanSpawnPoints =
-{
-    { 1776.27f, 1348.74f, 19.20f }, // spawn point for pumpkin shrine mob
-    { 1765.28f, 1347.46f, 17.55f }  // spawn point for smoke
-};
+std::vector<uint32> HeadlessHorsemanRandomLaughSound = { SOUNDID_MANIACAL_LAUGH, SOUNDID_MANIACAL_LAUGH_2, SOUNDID_MANIACAL_LAUGH_3 };
+Position const GOPumpkinSpawnPosition = { 1776.27f, 1348.74f, 20.4116f, 6.27281f };
+QuaternionData const GOPumpkinSpawnQuat = QuaternionData(0.0f, 0.0f, 0.00518764f, -0.999987f);
+Position const GOSoilSpawnPosition = { 1765.28f, 1347.46f, 17.5514f, 0.100363f };
+QuaternionData const GOSoilSpawnQuat = QuaternionData(0.0f, 0.0f, 0.0501604f, -0.998741f);
 
-//@todo Dear Lord, please someone have mercy and let this die soon
-static char const* HeadlessHorsemanInitialPlayerTexts[] =
+struct npc_headless_horseman_head : public PassiveAI
 {
-    "Horseman rise...",
-    "Your time is nigh...",
-    "You felt death once...",
-    "Now, know demise!"
-};
+    npc_headless_horseman_head(Creature* creature) : PassiveAI(creature), _phase(PHASE_1), _instance(creature->GetInstanceScript())
+    {
+        creature->SetDisplayId(me->GetCreatureTemplate()->Modelid1);
+    }
 
-struct npc_wisp_invis : public ScriptedAI
-{
-    npc_wisp_invis(Creature* creature) : ScriptedAI(creature), _timer(0s), _creatureType(0), _firstSpell(0), _secondSpell(0)
+    void JustAppeared() override
     {
-        creature->SetDisplayId(DISPLAYID_INVIS_WISP_INVISIBLE);
+        Initialize();
     }
 
-    void SetData(uint32 type, uint32 value) override
+    void Initialize()
```

Comment:
Initialize() is by convention supposed to be called from the ctor and from Reset() only to set the value of class members, without calling any AI function or doing any check on previous state as it might be uninitialized. Please pick another unused name if you have to do other stuff, i.e. casting spells.